Feature/p169 as3935 lightning detector

TD-er opened the pull request:
Also made quite a lot of changes to the PluginStats as logging lightning strikes can have days or even weeks between an event and thus the X-axis should reflect the time and not a sample number.

This then makes it harder to see details on different events which may have happened days or weeks apart.
So I also added a pan/zoom function:

![ZoomESPEasyPluginStatsCharts](https://github.com/letscontrolit/ESPEasy/assets/3751318/11b7da85-df74-41c1-824a-4c2f2af75196)

Zoom can be done by selecting a range or use the scroll wheel on the mouse.
Pan can be done by pressing [CTRL] along with left click and drag using the mouse.

On mobile zooming is done by pinching using 2 fingers and panning is just done by swiping over the image.

TD-er committed:
[AS3935] Add P169 Lightning Detector plugin

TD-er committed:
[AS3935] Improve calibration accuracy

TD-er committed:
[AS3935] Implement P169 code

TD-er committed:
[AS3935] Add option to measure all calibrated oscillators

TD-er committed:
[AS3935] Improve frequency measurement accuracy

TD-er committed:
[AS3935] Improve stability and calibration success rate

TD-er committed:
[AS3935] Make it compile for ESP8266 and clarify logs

tonhuisman commented on src/src/PluginStructs/P167_data_struct.h:
```diff
@@ -1,222 +1,227 @@
-//////////////////////////////////////////////////////////////////////////////////////////////////
-// P167 device class for IKEA Vindstyrka SEN54 temperature , humidity and air quality sensors 
```
Didn't expect this file to be in this PR?

TD-er committed:
[AS3935] Fix build on ESP32 IDF4.4

TD-er committed:
[AS3935MI] Get in sync with pending PR to original lib

TD-er committed:
[AS3935MI] Move frequency count code to original library code

TD-er committed:
[AS3935MI] Make frequency count "vanilla Arduino" for non-ESP SoC

TD-er committed:
[AS3935MI] Add selection for nr Samples, division, etc

TD-er committed:
[AS3935MI] Add '2nd best' check to calibration to retest

This way the best and 2nd best ant caps can be retested with longer test runs

TD-er committed:
[AS3935MI] Add event on lightning strike and count strikes per interval

TD-er committed:
[AS3935MI] Add total strikes counter, sensor stats and cleanup plugin

TD-er committed:
[AS3935MI] Cleanup settings & add chart to help tuning

TD-er committed:
[AS3935MI] Speedup LCO calibration + simplify configuration

TD-er committed:
Merge remote-tracking branch 'letscontrolit/mega' into feature/P169_AS3935_lightning_detector

tonhuisman commented on lib/AS3935MI/src/AS3935MI.h:
```diff
@@ -175,11 +176,11 @@ class AS3935MI
 
 	/*
-	@return true if disturbers are masked, false otherwise. */
+	@return true if disturbers are masked, false otherwisee. */
```
S&R side-effect?

TD-er replied:
Yep, should have clicked the match-whole-word box

TD-er committed:
[AS3935MI] Add timestamp to PluginStats

TD-er committed:
[AS3935] Fix compile issue

TD-er committed:
[ESP-IDF5.1] Update ESP-IDF SDK to fix HWCDC

TD-er committed:
[AS3935MI] Change lib function names to use camelCase

TD-er committed:
[ESP-IDF5.1] Fix crashes on LittleFS after factory reset

TD-er committed:
[ESP-IDF5.1] Update to latest ESP-IDF SDK code

TD-er committed:
[ESP-IDF 5.1] Exclude some OPI PSRAM speed improvements due to stability

TD-er committed:
[Json] Fix showing JSON including pluginstats data

TD-er committed:
[PluginStats] Compute average taking timestamps into account.

TD-er committed:
Merge remote-tracking branch 'letscontrolit/mega' into feature/P169_AS3935_lightning_detector

TD-er committed:
[Cleanup] Revert accidentally changed line endings

TD-er committed:
[PluginStats] Show average rate and total duration of collected stats

TD-er committed:
[PluginStats] Add nr of samples to chart legend

TD-er committed:
[PluginStats] Uncrustify/cleanup code

TD-er committed:
[PluginStats] Move stats to PSRAM when available

TD-er committed:
[AS3935MI] Rename interrupt enum to match style of other enums in lib

TD-er committed:
[Controller] Fail gracefully when almost out of memory

One often reported issue is that the ESPEasy node (especially ESP8266 ones) will crash when WiFi is unstable.

This could be due to memory suddenly filling up by the controller queue.
When creating a new item for the controller queue, there was no `(std::nothrow)` included on the `new` calls, so if running out of memory this might have thrown a `std::bad_alloc` exception, which are not caught and thus ESPEasy might crash on it.

TD-er committed:
[ESP-IDF5.1] Fix WiFi issues on ESP32-C6

TD-er committed:
[AS3935] Minor clean-up in code and user-interface

TD-er committed:
[P169] Start documenting AS3935 plugin

TD-er committed:
[AS3935MI] Get rid of 'magic' values in lib code

TD-er committed:
[AS3935] Fix "indoor/outdoor" setting + no longer disable on fail calib

TD-er committed:
[AS3935] Document calibration for P169 Lightning sensor

tonhuisman commented on docs/source/Plugin/P169_I2CAddressOptions.png:
This file is not in use for this plugin, can be removed

tonhuisman commented on docs/source/Plugin/P169_OutputRangeOptions.png:
Same for this file

tonhuisman commented on docs/source/Plugin/P169_commands.repl:
Same for this file

TD-er replied:
Not yet.... I just added calls for this

tonhuisman commented on docs/source/Plugin/P169_config_values.repl:
Same for this file

TD-er replied:
Not yet.... I just added calls for this

tonhuisman commented on docs/source/Plugin/P169.rst:
```diff
@@ -0,0 +1,335 @@
+I2C options
+^^^^^^^^^^^
+
+* **I2C Address**: The device supports 8 addresses, and by default comes configured for address ``0x03``.
```
8 => 3

TD-er committed:
[AS3935] Add commands to P169 Lightning Detector

TD-er committed:
[Build] Fix build due to static var with too generic name

TD-er committed:
[Docs] Remove unused files for P169

TD-er committed:
[Docs] Document getconfig and commands for P169 AS3935

TD-er committed:
[AS3935] Fix restart measurement after running calibration command

TD-er committed:
[AS3935] Set task values to output distance Far/Near and remove "energy"

TD-er committed:
[ESP-IDF5.1] update to latest ESP-IDF SDK & Arduino 3.0.0

TD-er committed:
[mDNS] Enable FEATURE_MDNS again for ESP-IDF5.1 builds

TD-er committed:
[AS3935MI] Add validation check for computeCalibratedFrequency

TD-er committed:
[AS3935MI] Make ISR attributes like IRAM_ATTR platform dependant.

TD-er committed:
[AS3935MI] Add auto gain algorithm to adapt to changes in noise envs.

TD-er committed:
[PluginStats] Update recorded timestamps when time is first set

TD-er committed:
[AS3935] Improve auto tuning

TD-er committed:
[PluginStats] Fix patching timestamps from before system time was set

TD-er committed:
[AS3935] Enable auto gain and reduce number of events

TD-er committed:
[mDNS] Disable mDNS again as it causes major stalls

TD-er committed:
[Charts] Add option to zoom/pan for PluginStats charts

TD-er committed:
[ESP-IDF5.x] Fix recent change of flush() to clear() in Arduino SDK

TD-er committed:
Revert "[ESP-IDF5.x] Fix recent change of flush() to clear() in Arduino SDK"

This reverts commit 5ee39150e29f20f8957c3dc8e5bbc851e19d94f6.

TD-er committed:
[ESP-IDF5.1] Revert Arduino PR 9453 WiFiClient - rename flush() to clear()

TD-er committed:
[ESP-IDF5.1] Update to latest platform build

TD-er committed:
[AS3935] Disable auto gain for now

TD-er committed:
Merge branch 'mega' into feature/P169_AS3935_lightning_detector

TD-er committed:
[Build] Fix merge issues.

TD-er committed:
[AS3935MI] Update to 1.3.5 with PRs from ESPEasy changes merged to lib

TD-er committed:
[System Load] Reduce system load running background tasks

TD-er committed:
[Cleanup] Speedup converting task values to String

TD-er committed:
[Cleanup] Further reduction of load

TD-er committed:
Merge remote-tracking branch 'letscontrolit/mega' into feature/P169_AS3935_lightning_detector

TD-er committed:
[Cleanup] Revert line ending to CRLF again (.h files)

TD-er committed:
[Cleanup] Revert changed line-ending back to CRLF (.cpp)

TD-er committed:
[Cleanup] Revert line endings from LF to CRLF

TD-er committed:
Merge remote-tracking branch 'letscontrolit/mega' into feature/P169_AS3935_lightning_detector

TD-er committed:
[Cleanup] Revert line ending to CRLF for PlatformIO ini files

TD-er committed:
[AS3935MI] Update increase/decrease functions return values in lib

TD-er committed:
[PluginStats] Decimate samples by not storing unchanged values

TD-er committed:
[Sonoff POR] Move Sonoff POW R1/R2 ESP8266 build back to core 2.7.4

Due to stability issues for Sonoff POW R2 which does reboot every 4 - 6 hours due to HW Watchdog when connected to modern AP.

TD-er committed:
[AS3935] Add selectable range for auto gain algorithm

TD-er committed:
[Build] Remove previously created __tmpfile.cpp files on ESP8266 build

TD-er committed:
[AS3935] Add check to see if RCO is still active on IRQ pin

TD-er committed:
[ESP-IDF5.x] Update to latest code incl. possible fix for WiFi issues

TD-er committed:
[Controller] Fix sending to controller

TD-er committed:
[Cleanup] Speedup all calls where task value name was used

A lot of extra work was done to determine the number of taskvalues, on each call.

Sending 4 task values to a MQTT controller (C005) for example took 20 msec on average, now about 10 msec.

tonhuisman commented on src/src/WebServer/CustomPage.cpp:
```diff
@@ -201,9 +201,9 @@ bool handle_custom(const String& path) {
             {
-              const String taskValueName = getTaskValueName(x, varNr);
+              const String taskValueName = Cache.getTaskDeviceValueName(x, varNr);
               if ((varNr < valueCount) &&
                   (!taskValueName.isEmpty()))
```
First part of this `if` condition is no longer needed 🤓

TD-er committed:
[Build] Fix build error on LIMIT_BUILD_SIZE builds

TD-er committed:
[ESP-IDF5.x] Update ESP-IDF/ESP32-Arduino  to latest commits

TD-er committed:
[ESPEasy p2p] Improve sharing tasks via p2p + add checksum to UDP packet

TD-er committed:
[ESPEasy p2p] Add timestamp to taskvalue data to counter network delays

TD-er committed:
[ESPEasy p2p] Fix sending taskinfo packet rejected by  wrong size check

TD-er committed:
[ESPEasy p2p] Make taskinfo length more flexible, remove trailing zeroes

TD-er committed:
[ESPEasy p2p] Add "Data Source" selection to request remote data feed

TD-er committed:
[Cleanup] Remove unused code

TD-er committed:
[MQTT] Fix some extreme timeouts when broker is unresponsive (#5075)

Fixes: #5075

TD-er committed:
[MQTT] Send keep-alive ping at 2/3rd of MQTT_KEEPALIVE

TD-er committed:
Merge branch 'mega' into feature/P169_AS3935_lightning_detector

TD-er committed:
[MQTT] Set timeout also for connecting to the MQTT broker

TD-er committed:
[Memory] Make special memory allocations more robust

TD-er committed:
[Controller] Add time spent in connecting to host to timing stats

TD-er committed:
[Cleanup] Rename ambiguously controller_number in controller code

TD-er committed:
[Controller] Add note on typical timeout for local vs. online networks

TD-er committed:
[PluginStats] Fix inconsistency with pluginstats for ADC plugin

TD-er committed:
[MQTT] Fix disconnect when broker sends keepalive ping before client

TD-er committed:
[ADC] Improve ADC curve fitting calibration accuracy

TD-er committed:
[VL53L0X] Fix ignoring sensor error values

TD-er committed:
[VL53L0X] Add async read option to library

TD-er committed:
[ESP-IDF5.x] Update to latest ESP-IDF/Arduino SDK

TD-er committed:
[WiFi] Add passive/active scan for ESP32

TD-er committed:
[Cleanup] Always use the same unit/note for settings, regardless the page

TD-er committed:
[ESP-IDF5.1] No longer support Ethernet on ESP32-C2

TD-er committed:
[Sysinfo] Add internal temperature to sysinfo (when supported)

TD-er committed:
[Build] Fix build due to missing #ifdef checks

TD-er committed:
[ESP32 temp] Fix int. temperature sensor not changing range

TD-er committed:
[WiFi] Fix WiFi not able to reconnect due to unprocessed disconnect

TD-er committed:
[VL53L0x] Make reading sensor async

TD-er committed:
[VL53L0X] Actually do read the sensor asynchronous.

TD-er committed:
[WiFi] Fix (re)connect to WiFi on ESP32-xx builds

TD-er committed:
[VL53L0X] Add filtering + make threshold more intuitive

TD-er committed:
[AS3935] Do not set IRQ pin to input when not configured (e.g. p2p share)

TD-er committed:
[Time] Simplify system time management + increase resolution to usec

TD-er committed:
[Time] Cleanup ESPEasy internal system time management

TD-er committed:
[VL53L0X] Cleanup code

TD-er committed:
[VL53L0X] Add dynamic filtering, allowing fast response on quick change

TD-er committed:
Merge branch 'mega' into feature/P169_AS3935_lightning_detector

TD-er committed:
[Time] Improve NTP accuracy and prevent lockups on incorrect data

TD-er committed:
[Cleanup] Fix wrong unit for TX power on web UI page
